File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ActivityLogTestWPCom.java
package org.wordpress.android.fluxc.release;

import org.greenrobot.eventbus.Subscribe;
import org.junit.After;
import org.junit.Test;
import org.wordpress.android.fluxc.TestUtils;
import org.wordpress.android.fluxc.action.ActivityAction;
import org.wordpress.android.fluxc.annotations.action.Action;
import org.wordpress.android.fluxc.generated.ActivityActionBuilder;
import org.wordpress.android.fluxc.model.activity.RewindStatusModel;
import org.wordpress.android.fluxc.store.ActivityLogStore;
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchActivitiesPayload;
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchedActivitiesPayload;
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchedRewindStatePayload;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;

public class ReleaseStack_ActivityLogTestWPCom extends ReleaseStack_WPComBase {
    @Inject ActivityLogStore mActivityLogStore;

    private CountDownLatch mBackup;
    private List<Action> mIncomingActions;

    @Override
    public void setUp() throws Exception {
        super.setUp();
        mReleaseStackAppComponent.inject(this);
        init();
        this.mBackup = this.mCountDownLatch;
        this.mIncomingActions = new ArrayList<>();
    }

    @Test
    public void testFetchActivities() throws InterruptedException {
        this.mCountDownLatch = new CountDownLatch(1);
        int numOfActivitiesRequested = 1;
        FetchActivitiesPayload payload = new FetchActivitiesPayload(sSite, numOfActivitiesRequested, 0);
        mActivityLogStore.onAction(ActivityActionBuilder.newFetchActivitiesAction(payload));
        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
        assertTrue(mIncomingActions.size() == 1);
        assertTrue(((FetchedActivitiesPayload) mIncomingActions.get(0).getPayload())
                           .getActivityLogModelRespons()
                           .size() == numOfActivitiesRequested);
    }

    @Test
    public void testFetchRewindState() throws InterruptedException {
        this.mCountDownLatch = new CountDownLatch(1);
        mActivityLogStore.onAction(ActivityActionBuilder
                .newFetchRewindStateAction(new ActivityLogStore.FetchRewindStatePayload(sSite, 10, 0)));
        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
        assertTrue(mIncomingActions.size() == 1);
        RewindStatusModel rewindStatusModelResponse =
                ((FetchedRewindStatePayload) mIncomingActions.get(0).getPayload()).getRewindStatusModelResponse();
        assertNotNull(rewindStatusModelResponse);
        assertNotNull(rewindStatusModelResponse.getState());
    }

    @After
    public void tearDown() throws Exception {
        this.mCountDownLatch = mBackup;
    }

    @SuppressWarnings("unused")
    @Subscribe
    public void onAction(Action action) {
        if (action.getType() instanceof ActivityAction) {
            if (mIncomingActions != null) {
                mIncomingActions.add(action);
                mCountDownLatch.countDown();
            }
        }
    }
}

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ActivityLogTestWPCom.kt
package org.wordpress.android.fluxc.release

import junit.framework.Assert.assertEquals
import org.greenrobot.eventbus.Subscribe
import org.junit.After
import org.junit.Test
import org.wordpress.android.fluxc.TestUtils
import org.wordpress.android.fluxc.action.ActivityAction
import org.wordpress.android.fluxc.annotations.action.Action
import org.wordpress.android.fluxc.generated.ActivityActionBuilder
import org.wordpress.android.fluxc.store.ActivityLogStore
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchActivitiesPayload
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchedActivitiesPayload
import org.wordpress.android.fluxc.store.ActivityLogStore.FetchedRewindStatePayload

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

import javax.inject.Inject

import junit.framework.Assert.assertNotNull
import junit.framework.Assert.assertTrue

class ReleaseStack_ActivityLogTestWPCom : ReleaseStack_WPComBase() {
    private val incomingActions: MutableList<Action<*>> = mutableListOf()
    @Inject lateinit var activityLogStore: ActivityLogStore
    private var mBackup: CountDownLatch? = null

    override fun setUp() {
        super.setUp()
        mReleaseStackAppComponent.inject(this)
        init()
        this.mBackup = this.mCountDownLatch
        this.incomingActions.clear()
    }

    @Test
    fun testFetchActivities() {
        this.mCountDownLatch = CountDownLatch(1)
        val numOfActivitiesRequested = 1
        val payload = FetchActivitiesPayload(ReleaseStack_WPComBase.sSite, numOfActivitiesRequested, 0)
        activityLogStore.onAction(ActivityActionBuilder.newFetchActivitiesAction(payload))
        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS.toLong(), TimeUnit.MILLISECONDS))
        assertTrue(incomingActions.size == 1)
        assertEquals((incomingActions[0].payload as FetchedActivitiesPayload)
                .activityLogModelRespons
                .size, numOfActivitiesRequested)
    }

    @Test
    fun testFetchRewindState() {
        this.mCountDownLatch = CountDownLatch(1)
        val payload = ActivityLogStore.FetchRewindStatePayload(ReleaseStack_WPComBase.sSite)
        activityLogStore.onAction(ActivityActionBuilder.newFetchRewindStateAction(payload))
        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS.toLong(), TimeUnit.MILLISECONDS))
        assertTrue(incomingActions.size == 1)
        val fetchedRewindStatePayload = incomingActions[0].payload as FetchedRewindStatePayload
        with(fetchedRewindStatePayload) {
            assertNotNull(this.rewindStatusModelResponse)
            assertNotNull(this.rewindStatusModelResponse?.state)
        }
    }

    @After
    fun tearDown() {
        this.mCountDownLatch = mBackup
    }

    @Subscribe
    fun onAction(action: Action<*>) {
        if (action.type is ActivityAction) {
            incomingActions.add(action)
            mCountDownLatch?.countDown()
        }
    }
}
